docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+
+query.id()          # get the unique identifier of the running query
+
+query.name()        # get the name of the auto-generated or user-specified name
+
+query.explain()   # print detailed explanations of the query
+
+query.stop()      # stop the query 
+
+query.awaitTermination()   # block until query is terminated, with stop() or with error
+
+query.exception()    # the exception if the query has been terminated with error
+
+query.souceStatus()  # progress information about data has been read from the input sources
+
+query.sinkStatus()   # progress information about data written to the output sink
+{% endhighlight %}
+
+</div>
+</div>
+
+You can start any number of queries in a single SparkSession. They will all be running concurrently sharing the cluster resources. You can use `sparkSession.streams()` to get the StreamingQueryManager that can be used to manage the currently active queries.
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+val spark: SparkSession = …
+
+spark.streams.active    // get the list of currently active streaming queries
+
+spark.streams.get(id)   // get a query object by its unique id
+
+spark.streams.awaitAnyTermination()   // block until any one of them terminates
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+{% highlight java %}
+SparkSession spark = …
+
+spark.streams().active()    // get the list of currently active streaming queries
+
+spark.streams().get(id)   // get a query object by its unique id
+
+spark.streams().awaitAnyTermination()   // block until any one of them terminates
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+
+{% highlight python %}
+spark = …  # spark session
+
+spark.streams().active()    # get the list of currently active streaming queries
+
+spark.streams().get(id)   # get a query object by its unique id
+
+spark.streams().awaitAnyTermination()   # block until any one of them terminates
+{% endhighlight %}
+
+</div>
+</div>
+
+Finally, for asynchronous monitoring of streaming queries, you can create and attach a StreamingQueryListener, which will give you regular callback-based updates when queries are started and terminated.
+
+## Recovering from Failures with Checkpointing 
+In case of a failure or intentional shutdown, you can recover the previous progress and state of a previous query, and continue where it left off. This is done using checkpointing and write ahead logs. You can configure query with a checkpoint location, and the query will save all the progress information (i.e. range of offsets processed in each trigger), and the running aggregates (e.g. word counts in the quick example) will be saved the checkpoint location. As of Spark 2.0, this checkpoint location has to be a path in a HDFS compatible file system, and can be set as an option in the DataStreamWriter when starting a query (see the previous section). 
```

Comment:
This is still fuzzy actually. I have to update in the future once we have the behavior more well defined. :(